### PR TITLE
[FIRRTL] Fix printf-encoded lint issues

### PR DESCRIPTION
**Pull request by seldridge**
Fix a bug when parsing certain types of printf-encoded
asserts ("assert:", "assume:", "cover:" and "assertNotX:") to not add an
unnecessary format string argument.  Instead, only add argumnets after
the first argument if any exist.  The latter case is extremely rare (to
use this from the Chisel APIs you would have to hand-craft a printf
operation and not use the built-in helper utilities).

Change Chisel asserts ("Assertion failed:") to always parse as immediate
assertions.  These assertions are expected to have format string
arguments and this avoids a catch-22 with VCS lint warnings.  A
concurrent assertion can be constructed to have a format string,
however, whichever way you craft it, VCS will give you a lint warning:

Ill-advised, unsampled option A:

  assert_1: assert property (@(posedge clock) condition)
    else $error("The value of 'foo' is %d", foo);

Advised, sampled optiona B:

  assert_2: assert property (@(posedge clock) condition)
    else $error("The value of 'bar' is %d", $sampled(bar));

For more information, see Section 16.9.3 of the SystemVerilog
specification (which is from where the above code snippet is adapted).

Signed-off-by: Schuyler Eldridge <schuyler.eldridge@sifive.com>

### CODE REVIEW

File: lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp [hunk 1]
```diff
@@ -413,24 +413,40 @@ ParseResult circt::firrtl::foldWhenEncodedVerifOp(PrintFOp printOp) {
 
     // TODO: Sanitize the label by replacing whitespace with "_" as done in the
     // Scala impl of ExtractTestCode.
+    ValueRange args;
+    if (printOp.operands().size())
+      args = printOp.operands().drop_front();
+    if (args.size())
+      printOp.emitWarning()
+          << "printf-encoded assertion has format string arguments which may "
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp:421]: This is going to trigger fairly often, yes?

**Review by darthscsi — APPROVED** [review]
LGTM

File: lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp (hunk 1 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp:421]: I don't think so. This has to be an assert that is printf-encoded, uses the `assert:`/`assume:`/`cover:`/`assertNotX:`, _and the user has not used the builtin helper utilities to generate this_.

Builtin utilities lock this down and won't generate `printf` with format string arguments, e.g., `assert:` comes from here:
  - https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/formal/FormalUtils.scala#L72

There are a few usages of raw `printf`-encoded verification statements that do not use the builtin utilities, but I didn't see any that have format string arguments.

tl;dr: seeing this warning should be extremely rare.

File: lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp [hunk 2]
```diff
@@ -413,24 +413,40 @@ ParseResult circt::firrtl::foldWhenEncodedVerifOp(PrintFOp printOp) {
...
+    // if these are emitted as concurrent assertions:
+    //   1. VCS warns about non-sampled values used in the pass/fail statements
+    //      of concurrent assertions.
+    //   2. VCS warns if you sample the values ($sample) indicating that the
+    //      assertion has side effects.
   case VerifFlavor::ChiselAssert: {
     builder.create<AssertOp>(
         printOp.clock(), builder.create<NotPrimOp>(whenStmt.condition()),
-        printOp.cond(), fmt, printOp.operands(), "chisel3_builtin", true);
+        printOp.cond(), fmt, printOp.operands(), "chisel3_builtin", false);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParserAsserts.cpp:449]: Note: we can change this back to concurrent asserts once https://github.com/llvm/circt/pull/2549 lands.
